### PR TITLE
Update support for parameterised volumes

### PR DESCRIPTION
**Pull request by ihrivnac**
Updates in the previously added code
- Added 'volumes' argument in IFactory::CreateParameterisedPlacement
  and IPlacement::ParameterisedPlacementData() and adapted implementation
  for these changes
  - this avoids relying on naming convention chosen for
    paremeterised volumes elements that may clash with existing volume
    names
- Renamed local variables so that they start with lowercase letter
- Improved debug output

Updated test for parameterised volumes:

- Added new test Placements3 in test suites
- Split parameterisation class in two for better clarity
- Use the same names for logical and physical volumes in the added test
  (this gives better matchin in the run test output)
- Renamed local variables so that they start with lowercase letter
  and do not contain '_'

### CODE REVIEW

**Comment by ihrivnac** [comment]
Hi @hpekeler ,

Could you, please, test this update?

As described above, your implementation has some weakness (relying on volumes names) that is improved in this update. On my side your test gives the same results as before.

To update your repository with this PR:
```
git checkout -b ihrivnac-update-parameterised master
git pull https://github.com/ihrivnac/vgm.git update-parameterised

```

Best regards,

**Comment by hpekeler** [comment]
Sure i can check it out!

During compiling i get the following warning: 
```
/localhome/pekeler/Tools/vgm/packages/Geant4GM/source/volumes/Factory.cxx:477:13: warning: init-statement in selection statements only available with ‘-std=c++17’ or ‘-std=gnu++17’
  477 |         if (auto itv = localSolidMaterialStore.find(std::make_pair(pSolid, pMat));
```
I would advise to either change this statement or require c++17. The last one is probably not the best idea?

**Comment by ihrivnac** [comment]
> I would advise to either change this statement or require c++17.

Thank you for pointing at this; Geant4 already requires c++17, but as Root can be built with c++11, we should better keep a possibility not to require c++17. I will change this code.

**Comment by hpekeler** [comment]
thanks for this update. I indeed hated that i did some string compare. Introducing this new global store makes much sense and looks way better. From reading the code i find it perfectly fine. I would change one line of code which i will link in the next comment.  
I also tested the code with my geometries and everything looks fine.

File: packages/RootGM/source/volumes/Factory.cxx [hunk 1]
```diff
@@ -1026,13 +1027,19 @@ VGM::IPlacement* RootGM::Factory::CreateParameterisedPlacement(
...
+    std::cerr << "    RootGM::Factory::CreateParameterisedPlacement:"
+              << std::endl;
+    std::cerr << "    Transformations and Volumes vector sizes cannot be different!"
+              << std::endl;
+    std::cerr << "*** Error: Aborting execution  ***" << std::endl;
+    exit(1);
+  }
+
+  for (size_t i = 0; i < transforms.size(); ++i) {
+    std::cout << i << " " << volumes[i]->Name() << " " << transforms[i] << std::endl;
```
Comment (hpekeler) [inline packages/RootGM/source/volumes/Factory.cxx]: put in debug scope?

Comment (ihrivnac) [reply packages/RootGM/source/volumes/Factory.cxx]: This can be removed

Comment (hpekeler) [reply packages/RootGM/source/volumes/Factory.cxx]: Also fine

**Comment by ihrivnac** [comment]
Thank you for testing; I will merge this and prepare a tag

**Comment by hpekeler** [comment]
Thank you very much :+1: :-)

**Pull request merged**